### PR TITLE
MSC2290: Separate Endpoints for Threepid Binding

### PR DESCRIPTION
**Pull request by anoadragon453**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/anoa/msc_separate_hs_api/proposals/2290-separate-threepid-bind-hs.md)

~~TODO: Backwards compatibility is still being figured out.~~

Implementation: https://github.com/matrix-org/synapse/pull/6043

### CODE REVIEW

File: proposals/2290-separate-threepid-bind-hs.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
+# Separate Endpoints for Binding Threepids
+
+On the Client Server API there is currently a single endpoint for binding a
+threepid (an email or a phone number): [POST
+/_matrix/client/r0/account/3pid](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-3pid).
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: I believe I've made it easier to read now? Please check.

[the remark this answers is not in the document]
Comment (turt2live) [reply proposals/2290-separate-threepid-bind-hs.md]: yup - thank you!

File: proposals/2290-separate-threepid-bind-hs.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+{
+    "sid": "abc123987",
+    "client_secret": "don'tT3ll"
+}
+```
+
+The threepid will then be bound to the user's account.
+
+The achieve the above flows, some changes need to be made to existing
```
Comment (ara4n) [inline proposals/2290-separate-threepid-bind-hs.md]: To achieve

File: proposals/2290-separate-threepid-bind-hs.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        "client_secret": "don'tT3ll"
+    }
+}
+```
+
+The homeserver will then make a bind request to the specified identity server
+on behalf of the user. The homeserver will record if the bind was successful
+and notify the user.
+
+The threepid has now been binded on the user's requested identity server
```
Comment (ara4n) [inline proposals/2290-separate-threepid-bind-hs.md]:
```suggestion
The threepid has now been bound on the user's requested identity server
```

File: proposals/2290-separate-threepid-bind-hs.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    "client_secret": "don'tT3ll",
+    "email": "bob@example.com",
+    "send_attempt": 1,
+}
+```
+
+Once an email has been sent, the user clicks the link in the email, which
+notifies the homeserver that the threepid has been verified.
+
+The client then sends a request to the old endpoint on the homeserver to bind
```
Comment (ara4n) [inline proposals/2290-separate-threepid-bind-hs.md]: why do we call this 'the old endpoint'?

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: As we're deprecating it, but I can be more explicit.

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Typo, should be 'other'.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+about hundreds of fake binds to a user's account).
+
+This MSC obseletes
+[MSC2229](https://github.com/matrix-org/matrix-doc/pull/2229), which dealt
+with changing the rules of the `bind` flag on the original endpoint. Since
+that endpoint is being deprecated, the MSC is no longer relevant.
+
+## Proposal
+
+Two new endpoints will be added to the Client Server API: `POST
```
Comment (ara4n) [inline proposals/2290-separate-threepid-bind-hs.md]: I think the proposal would be much easier to grok if the first paragraph explained what the two endpoints actually do.  So `/bind` publishes the 3PID mapping on an IS, and `/add` adds an administrative 3PID to the HS?

**Comment by ara4n** [comment]
lgtm other than my q's. let's FCP it?

File: proposals/2290-separate-threepid-bind-hs.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
+# Separate Endpoints for Binding Threepids
+
+On the Client Server API there is currently a single endpoint for binding a
+threepid (an email or a phone number): [POST
+/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid).
+Depending on whether the `bind` flag is `true` or `false`, the threepid will
+be bound to either a user's account on the homeserver, or both the homeserver
```
Comment (jryans) [inline proposals/2290-separate-threepid-bind-hs.md]: In an attempt to preserve some shreds on sanity with 3PIDs, I have been using the verb "bind" _only_ for the case of identity server. For adding a 3PID to an HS account, I use the verb "add". Using separate verbs in this way makes the concepts easier to distinguish.

I think this MSC would be easier to read if it followed the same convention, as opposed to have two meanings of binding depending on the service.

Comment (ara4n) [reply proposals/2290-separate-threepid-bind-hs.md]: definitely agreed. (plus we should define that terminology up front)

File: proposals/2290-separate-threepid-bind-hs.md (hunk 5 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Yep, and ok can do. I suppose that is the idea of [BLUF](https://news.ycombinator.com/item?id=20964907).

File: proposals/2290-separate-threepid-bind-hs.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+Synapse is soon to lose this whitelist however, as part of lessening the
+reliance of homeservers on identity servers. This cannot be done while the
+homeserver is still trusting an identity server for validation of threepids.
+If the endpoints are split, the homeserver will handle the validation of
+threepids being added to user accounts, and identity servers will validate
+threepids being added to their own database.
+
+To solve this problem, we propose adding two new endpoints. One that is only
+used for binding to user's account, and another that is only for binding to
```
Comment (jryans) [inline proposals/2290-separate-threepid-bind-hs.md]: Again, this overview would be simpler to follow using the separate verbs "add" and "bind" for the actions that take place. Conveniently, these verbs also match the API paths, so that should make it even more straightforward.

**Comment by anoadragon453** [comment]
`bind` and `add` now mean different things.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+/account/3pid/email/requestToken](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid-email-requesttoken)
+and [POST
+/account/3pid/msisdn/requestToken](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid-msisdn-requesttoken)
+endpoints, as these endpoints are now only intended for the homeserver to
+send validation requests from.
+
+Additionally, the [POST
+/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid)
+endpoint is deprecated as the two new endpoints replace its functionality.
+The `bind` endpoint will also be removed, with the endpoint functioning as if
```
Comment (jryans) [inline proposals/2290-separate-threepid-bind-hs.md]: What is that `bind` endpoint you are removing? (I thought this MSC was adding the bind endpoint...)

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: whoops, meant "parameter", not endpoint. Thanks.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
+# Separate Endpoints for Binding Threepids
+
+On the Client Server API there is currently a single endpoint for binding a
+threepid (an email or a phone number): [POST
+/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid).
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: please link to a specific version if at all possible. If the endpoint can't be linked to a specific version, mention that it's the endpoint scheduled for r0.6.0

File: proposals/2290-separate-threepid-bind-hs.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+to supply a malicious identity server that would immediately answer "yes" to
+any threepid validation, then the user could add any threepid to their
+account on the homeserver (which is likely not something homeserver admins want).
+
+To be clear, this is not a long-standing security issue. It is not a problem
+in any released version of Synapse, as Synapse keeps a list of "trusted
+identity servers" that acts a whitelist for what identity servers a user can
+specify.
+
+Synapse is soon to lose this whitelist however, as part of lessening the
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: while great that Synapse wants to separate itself from identity servers, it's not a problem for the spec to deal with. Please present the feature without the reliance on Synapse - we wouldn't merge a feature that was implementation-specific because that implementation wanted it.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+reliance of homeservers on identity servers. This cannot be done while the
+homeserver is still trusting an identity server for validation of threepids.
+If the endpoints are split, the homeserver will handle the validation of
+threepids being added to user accounts, and identity servers will validate
+threepids being bound to themselves.
+
+To solve this problem, we propose adding two new endpoints. `POST
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: we? 
```suggestion
To solve this problem, it is proposed to add two new endpoints. `POST
```

File: proposals/2290-separate-threepid-bind-hs.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+This MSC obseletes
+[MSC2229](https://github.com/matrix-org/matrix-doc/pull/2229), which dealt
+with changing the rules of the `bind` flag on the original endpoint. Since
+that endpoint is being deprecated, the MSC is no longer relevant.
+
+## Proposal
+
+Two new endpoints will be added to the Client Server API: `POST
+/account/3pid/bind` and `POST /account/3pid/add`. Both will require
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: auth + can be rate limited presumably. I know we've done a bad job of determining what can be rate limited thus far, but at some point we should stop being bad - this MSC is very much an arbitrary point for me to raise it.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+Once an email has been sent, the user clicks the link in the email, which
+notifies the identity server that the email has been verified.
+
+Next, the client completes the bind by calling the new endpoint on the homeserver:
+
+```
+POST https://home.server/_matrix/client/r0/account/3pid/bind
+
+{
+    "three_pid_creds": {
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: let's fix https://github.com/matrix-org/matrix-doc/issues/2220 while we're here and converge on exactly one. I don't think we need to be terse here and can say "threepid_credentials"

Comment (turt2live) [reply proposals/2290-separate-threepid-bind-hs.md]: or better yet just not wrap the object

File: proposals/2290-separate-threepid-bind-hs.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+Additionally, the [POST
+/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid)
+endpoint is deprecated as the two new endpoints replace its functionality.
+The `bind` parameter will also be removed, with the endpoint functioning as if
+`bind` was `false`. Allowing an endpoint to add a threepid to both the
+identity server and homeserver at the same time requires one to trust the
+other, which is the exact behaviour we're trying to eliminate. Doing this
+also helps backward compatibility, as explained below.
+
+This MSC also requests that the text "It is imperative that the homeserver
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: MSCs are demands by nature (proposals go forward with recommendations leading to action, not listing things that don't require action) - if this is meant to be actionable, please phrase it as such. My recommendation is:

> The text "It is imperative...that it trusts" is to be removed from all parts of the spec, as the homeserver should no longer need to trust any identity servers.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+/account/3pid/email/requestToken](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid-email-requesttoken)
+and [POST
+/account/3pid/msisdn/requestToken](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid-msisdn-requesttoken)
+endpoints, as these endpoints are now only intended for the homeserver to
+send validation requests from.
+
+Additionally, the [POST
+/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid)
+endpoint is deprecated as the two new endpoints replace its functionality.
+The `bind` parameter will also be removed, with the endpoint functioning as if
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]:
```suggestion
The `bind` parameter is to be removed, with the endpoint functioning as if
```

File: proposals/2290-separate-threepid-bind-hs.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+longer need to trust any identity servers.
+
+## Tradeoffs
+
+It may be possible to reduce the two calls per flow into a single endpoint,
+but the current asynchronous approach makes it easy for a client to send a
+request, go offline, have the threepid be validated, and then come online
+again to finalize the validation afterwards.
+
+## Backwards compatibility
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: Does this need an unstable feature flag so that clients wiring up these new APIs can detect their presence?

File: proposals/2290-separate-threepid-bind-hs.md [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+receiving a HTTP `404` error code, should either attempt to use
+`/account/3pid` with the `bind` parameter or give up, at their discretion.
+
+## Security considerations
+
+Reducing the homeserver's trust in identity servers should be a boost to
+security and improve decentralisation in the Matrix ecosystem to boot.
+
+Caution should be taken for homeserver developers to be sure not to continue
+to use user-provided identity servers for any sensitive tasks once it removes
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: I have no idea what this entails - what sensitive tasks is this referencing? why am I supposed to be careful but allow any ID server to go through?

File: proposals/2290-separate-threepid-bind-hs.md [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+the concept of a trusted identity server list.
+
+## Conclusion
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: I feel like this whole section can be removed from this MSC (and the template, for that matter)

File: proposals/2290-separate-threepid-bind-hs.md [hunk 19]
```diff
@@ -0,0 +1,206 @@
+# Separate Endpoints for Binding Threepids
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: The intro duplicates the proposal and is a bit verbose - it can be shortened by not explaining what the proposal is about to do and just setting the scene for why the changes are being made. 

The intro is already desynchronized from the proposal itself: it does not mention that `bind` is disappearing.

File: proposals/2229-rebind-existing-3pid.md [hunk 20]
```diff
@@ -1,5 +1,7 @@
 # Allowing 3PID Owners to Rebind
 
+## Note: This MSC has been made obselete by MSC2290.
```
Comment (turt2live) [inline proposals/2229-rebind-existing-3pid.md]: I'm failing to see how MSC2229 is insufficient for this problem. It's great that the MSC is called out as obsolete and how it is obsolete, however the explanation for why 2229 is sub-par is missing.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 16 above)
Comment (jryans) [reply proposals/2290-separate-threepid-bind-hs.md]: Hmm, I suppose an unstable flag might make the client work slightly simpler... Without a flag, you would need to probe the new endpoints and check for 404s to detect support. If it were just a single API call, 404 detection could be okay, but in this case, these APIs are part of a multiple step flow of requesting tokens, sending to a service, etc. and the new endpoints are not the first step, so it means the 404 detection would be an extra check before starting the entire flow.

File: proposals/2229-rebind-existing-3pid.md (hunk 20 above)
Comment (jryans) [reply proposals/2229-rebind-existing-3pid.md]: Some of the advantages over MSC2229 include:

* Splitting the APIs makes the flow simpler to reason about
* Different services will perform the 3PID ownership verification: for `add`, it would be the HS or a trusted delegate; for `bind`, it would be the user supplied IS

Comment (anoadragon453) [reply proposals/2229-rebind-existing-3pid.md]: I've clarified on the doc.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+}
+```
+
+Once an email has been sent, the user clicks the link in the email, which
+notifies the identity server that the email has been verified.
+
+Next, the client completes the bind by calling the new endpoint on the homeserver:
+
+```
+POST https://home.server/_matrix/client/r0/account/3pid/bind
```
Comment (jryans) [inline proposals/2290-separate-threepid-bind-hs.md]: Does the new HS `bind` endpoint require the 3PID to already have been added to the HS account previously? (I assume the answer is no, but the MSC doesn't appear to specify.)

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: No, as the IS doesn't allow that. I'll add it in though, thanks.

**Comment by jryans** [comment]
I have updated the [3PID API flow diagrams](https://gist.github.com/jryans/839a09bf0c5a70e2f36ed990d50ed928) to account for this MSC.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+The concept of this whitelist is being removed in this MSC however, as part
+of lessening the reliance of homeservers on identity servers. This cannot be
+done while the homeserver is still trusting an identity server for validation
+of threepids. If the endpoints are split, the homeserver will handle the
+validation of threepids being added to user accounts, and identity servers
+will validate threepids being bound to themselves.
+
+## Proposal
+
+To solve this problem, two new endpoints will be added to the Client Server
```
Comment (richvdh) [inline proposals/2290-separate-threepid-bind-hs.md]: I might be missing it, but: is it ever spelt out what the HS is meant to do when these endpoints are hit? It's kinda implied by reference to the existing endpoint, but it seems a bit unclear to me.

Comment (richvdh) [inline proposals/2290-separate-threepid-bind-hs.md]: Relatedly: my main question is how hitting /bind on the HS differs from hitting the IS directly.

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: The HS will just proxy this request. However, it will remember that the bind occurred, and when it comes time to unbind all of a user's threepids (during account deactivation or when they explicitly request it), then the homeserver is aware of what binds occurred where and can carry out the unbinds on the user's behalf.

But other than that it is simply just proxying the request.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 16 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: I've switched it to use a new unstable flag, `m.separate_add_and_bind`.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 22 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: I've stated after each call what the accompanying server should have done internally. Hopefully that makes things clearer.

**Comment by anoadragon453** [comment]
Is this good to call FCP on do people think?

File: proposals/2290-separate-threepid-bind-hs.md [hunk 23]
```diff
@@ -191,16 +188,8 @@ receiving a HTTP `404` error code, should either attempt to use
...
-This MSC helps to minimize the homeserver's trust in an identity server even
-further to the point where it is only used for binding addresses for lookup -
-which was the original intention of identity servers to begin with.
-
-Additionally, by clearly separating the original threepid endpoint into two
-endpoints that each have a clear intention, the concept of attaching
-threepids to a Matrix user becomes a lot easier to reason about.
+Some endpoints of the Client Server API allow a user to provide an
+`id_server` parameter. Caution should be taken for homeserver developers to
+stop using these user-provided identity servers for any sensitive tasks, such
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: where possible, presumably. If it is flagged as required, there's not much the homeserver can do. If it is supposed to be optional in more places, please open another MSC.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 24]
```diff
@@ -133,8 +133,8 @@ POST https://home.server/_matrix/client/r0/account/3pid/add
 The threepid has now been added to the user's account.
 
 To achieve the above flows, some changes need to be made to existing
-endpoints. This MSC requests that the `id_server` and `id_access_token`
-parameters be removed from the Client-Server API's [POST
+endpoints. The `id_server` and `id_access_token` parameters are to be removed
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: removed or made optional? How does one proxy the calls to an IS if the fields aren't present?

File: proposals/2290-separate-threepid-bind-hs.md (hunk 23 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Yeah, where possible.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 24 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Err, yes, made optional as the spec still allows the homeserver to proxy requests to identity servers specified by the client.

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Actually, these parameters should be removed as it's dangerous to remove the requirement for a trusted identity server list while still allowing a user-provided `id_server` parameter. Really we should be removing `id_server` and `id_access_token` from all `/requestToken` endpoints. I'll add that in.

**Comment by anoadragon453** [comment]
Calling FCP now as I think this MSC is asking everything we want it to, though the addition in https://github.com/matrix-org/matrix-doc/pull/2290/commits/bd64ffc442127774a9f34631a3960be43dd975d8 might require some back and forth explanation.

@mscbot fcp merge

**Comment by anoadragon453** [comment]
> Calling FCP now as I think this MSC is asking everything we want it to, though the addition in bd64ffc might require some back and forth explanation.

Essentially if we remove the need for a identity server whitelist, we're opening ourselves up to the attack described in this MSC. Part of the solution is removing the user's ability to provide the homeserver with an arbitrary domain to use for threepid validation. Since homeservers should be handling all the validation themselves, then the user does not need to provide an identity server to handle those tasks. Thus we should remove the `id_server` and `id_access_token` parameters in all of the `/requestToken` endpoints.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 25]
```diff
@@ -145,27 +145,33 @@ given ID and client_secret was validated, and if so adds the threepid to the
...
-also helps backward compatibility, as explained below.
-
-The text "It is imperative that the homeserver keep a list of trusted
-Identity Servers and only proxies to those that it trusts." is to be removed
-from all parts of the spec, as the homeserver should no longer need to trust
-any identity servers.
+also helps backward compatibility, as explained in [Backwards
+compatibility](#backwards-compatibility).
+
+The `id_server` and `id_access_token` parameters are to be removed
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: So this obsoletes [MSC2263](https://github.com/matrix-org/matrix-doc/pull/2263) too then? :/

They were made optional in 2263 for the reason of maintaining backwards compatibility with clients. Removing them does not maintain that backwards compatibility. In fact, 2263 even introduces an error code to let the HS make the decision about whether or not it cares through introduction of an error code.

This is duplicated effort.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 26]
```diff
@@ -145,27 +145,33 @@ given ID and client_secret was validated, and if so adds the threepid to the
...
+* [POST /account/password/email/requestToken](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-password-email-requesttoken)
+* [POST /account/password/msisdn/requestToken](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-password-msisdn-requesttoken)
+
+Either the homeserver itself or a service that the homeserver delegates to
+should be handling the sending of validation messages, not a user-provided
+server. Any mention of the homeserver being able to proxy to an identity
+server in the above endpoint descriptions, as well as the text "It is
+imperative that the homeserver keep a list of trusted Identity Servers and
+only proxies to those that it trusts." is to be removed from all parts of the
+spec, as the homeserver should no longer need to trust any identity servers.
```
Comment (turt2live) [inline proposals/2290-separate-threepid-bind-hs.md]: There's still an element of trust, as highlighted by yourself here: https://github.com/matrix-org/matrix-doc/pull/2263/files#r319536914

**Comment by turt2live** [comment]
@mscbot concern This MSC is overreaching.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 27]
```diff
@@ -0,0 +1,215 @@
...
+```
+
+The identity server must send an email to the specified address, including a
+link to a URL on the identity server which will accept the validation session
+ID, the given client_secret, and a randomly-generated token.
+
+Once an email has been sent, the user clicks the link in the email, which
+notifies the identity server that the email has been verified.
+
+Next, the client completes the bind by calling the new endpoint on the homeserver:
```
Comment (KitsuneRal) [inline proposals/2290-separate-threepid-bind-hs.md]: That "next" thing can (still) be triggered by `next_link` URL, right? Or am I outdated and we have a new way of passing the baton from IS back to the client application without awkward "Click here once you confirmed the email" buttons?

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: No, we still have `next_link` shenanigans. It's specified in the `/requestToken` step, and homeservers should redirect users if it was specified as part of the validation session.

I can add a note for that.

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Actually, that part is handled on `submit_token`, which is covered by [MSC2263](https://github.com/matrix-org/matrix-doc/pull/2263). This MSC only deals with changing the `/account/3pid` endpoint, so I don't think `next_link` needs to be mentioned.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 26 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Mm, Homeserver can get around this in implementation though (as we have done in Synapse) by blocking attempts to access private IP address ranges.

Comment (turt2live) [reply proposals/2290-separate-threepid-bind-hs.md]: to be fair, keeping a list of servers that *aren't* trusted is by proxy keeping a list of servers that *are* trusted. However, yea: this is a bit of a pointless argument

File: proposals/2290-separate-threepid-bind-hs.md (hunk 25 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: We can leave them as optional parameters but homeserver should ignore them else they are opening themselves up to malicious identity servers.

And then at that point there's not much point in keeping them as optional parameters versus removing them.

Comment (turt2live) [reply proposals/2290-separate-threepid-bind-hs.md]: By removing them, it becomes impossible for `/bind` to work. The client will have no way of contacting an identity server through the homeserver, so the homeserver might end up rejecting `/bind`. If there's no way to proxy the calls to the IS through the HS, there's no point in `/bind` and we should just remove the `bind` flag from `/account/3pid` instead of splitting the endpoint.

I'd rather we keep the fields as optional and put a red box saying "you should abuse M_SERVER_NOT_TRUSTED at every given opportunity" until the deprecation period has passed and they get removed (as they are deprecated by MSC2263).

**Comment by dbkr** [comment]
lgtm modulo travis's concern of conflicting with MSC2263.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 25 above)
Comment (dbkr) [reply proposals/2290-separate-threepid-bind-hs.md]: So, iiuc the problem is:
 * We don't want to support `id_server` in synapse because we want to ditch the concept of trusted ID servers
 * This isn't compatible with MSC2263 because that requires that the endpoints still work when given `id_server`

I think, as long as the requests still work in synapse with an `id_server` (modulo the fact they won't bind, but that's covered by this MSC), that's fine? The HS will send a token (not via the IS specified) but the right thing will still happen.

So I would suggest maybe keeping them optional for now?

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: Keeping them optional but returning `M_SERVER_NOT_TRUSTED` works I suppose.

Comment (turt2live) [reply proposals/2290-separate-threepid-bind-hs.md]: 2263 was specifically written to allow Synapse to not support `id_server` ftr: it can just throw `M_SERVER_NOT_TRUSTED` for everything.

File: proposals/2229-rebind-existing-3pid.md [hunk 28]
```diff
@@ -1,5 +1,16 @@
...
```
Comment (uhoreg) [inline proposals/2229-rebind-existing-3pid.md]:
```suggestion
## Note: This MSC has been made obsolete by MSC2290.
```

File: proposals/2290-separate-threepid-bind-hs.md [hunk 29]
```diff
@@ -0,0 +1,215 @@
+# Separate Endpoints for Binding Threepids
+
+*Note: This MSC obseletes
```
Comment (uhoreg) [inline proposals/2290-separate-threepid-bind-hs.md]:
```suggestion
*Note: This MSC obsoletes
```

**Comment by uhoreg** [comment]
@mscbot concern /add should require UI Auth

**Comment by anoadragon453** [comment]
> @mscbot concern /add should require UI Auth

I've added a requirement for `/3pid/account/add` to have UIAA now. I don't think it's necessary for `/bind` as that can be done out of band with a given identity server anyways.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 25 above)
Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: We can't have homeservers throw `M_SERVER_NOT_TRUSTED` else old clients trying to send `id_server` will get confused when they get this error.

But new servers can just ignore the parameter.

**Comment by anoadragon453** [comment]
> @mscbot concern This MSC is overreaching.

I've taken out the bit where `id_server` and `id_access_token` are removed from `/requestToken` endpoints.

**Comment by uhoreg** [comment]
@mscbot resolve /add should require UI Auth

**Comment by turt2live** [comment]
@mscbot resolve This MSC is overreaching.

File: proposals/2290-separate-threepid-bind-hs.md (hunk 25 above)
Comment (turt2live) [reply proposals/2290-separate-threepid-bind-hs.md]: I'd still expect that servers throw the error instead of ignoring it, as ignoring it is against the spec.

It'd be no different than a 500 error in the eyes of the client (which it should be handling because `/requestToken` is certainly in the top 10 for most error-prone endpoints).

Comment (jryans) [reply proposals/2290-separate-threepid-bind-hs.md]: From a client perspective, we'd like old Riots to still be able add 3PIDs to the HS account. 

If `requestToken` begins throwing an error when `id_server` is included, those clients will no longer have any path to add a 3PID. The old Riots would have had known to special case and ignore `M_SERVER_NOT_TRUSTED`, which did not exist in the past.

File: proposals/2290-separate-threepid-bind-hs.md [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+of lessening the reliance of homeservers on identity servers. This cannot be
+done while the homeserver is still trusting an identity server for validation
+of threepids. If the endpoints are split, the homeserver will handle the
+validation of threepids being added to user accounts, and identity servers
+will validate threepids being bound to themselves.
+
+## Proposal
+
+To solve this problem, two new endpoints will be added to the Client Server
+API: `POST /account/3pid/bind` and `POST /account/3pid/add`. Binding to an
```
Comment (erikjohnston) [inline proposals/2290-separate-threepid-bind-hs.md]: Its possibly a bit late in the day, but was there any discussion about using slightly more informative names? `POST /account/3pid/bind_to_is` or something? I had to look up the difference between `bind` and `add` when skimming this.

Comment (anoadragon453) [reply proposals/2290-separate-threepid-bind-hs.md]: We've used the `bind` vs. `add` terminology quite extensively for this project and it seems to have gone down well as soon as you're aware of the difference. So perhaps it's just a case of defining the terms in the spec?

File: proposals/2290-separate-threepid-bind-hs.md (hunk 25 above)
Comment (turt2live) [reply proposals/2290-separate-threepid-bind-hs.md]: @jryans sounds like the server shouldn't be throwing the error then and instead handle the `id_server` as deprecated (as proposed by #2263). If the server doesn't want to handle it, it is no different from it throwing a 500 error for a misconfigured email server.

It would be intentional that adding 3PIDs be blocked - it's an error related to something the server is not comfortable with doing.

This all should have come out on MSC2263, not here, as well. I'd still have the same opinion, but it's off-topic for this MSC (now).

**Comment by turt2live** [comment]
@anoadragon453 does this have an implementation already?

**Comment by anoadragon453** [comment]
Oop, sorry, yes it does. I'll put them in the top comment.

**Comment by turt2live** [comment]
Cool, for my sanity: Implementation at https://github.com/matrix-org/synapse/pull/6043

**Comment by turt2live** [comment]
Spec PR: https://github.com/matrix-org/matrix-doc/pull/2328

**Comment by turt2live** [comment]
merged 🎉
